### PR TITLE
refactor(bigtable): use variant<status, T> not StatusOr<optional<T>>

### PR DESCRIPTION
**Pull request by dbolduc**
refactor `RowReader` / `RowReaderIterator` to deal in `absl::variant<Status, Row>`s instead of `StatusOr<absl::optional<Row>>`s. This is more consistent with the interface `StreamRange` expects. In a subsequent PR I will probably implement RowReader in terms of a StreamRange (which will allow me to delete `bigtable/internal/rowreader_iterator.{h,cc}`)

 | previous type: `StatusOr<absl::optional<Row>>` | new type: `absl::variant<Status, Row>` |
|-|-|
| has `Row` | has `Row`|
| bad Status | has bad Status |
| has `absl::nullopt` | has OK Status|

<hr />

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/8948)

### CODE REVIEW

File: google/cloud/bigtable/internal/legacy_row_reader_impl.cc [hunk 1]
```diff
@@ -105,25 +105,28 @@ bool LegacyRowReaderImpl::NextChunk() {
...
-    grpc::Status status = AdvanceOrFail(row);
+    auto variant = AdvanceOrFail();
+    if (absl::holds_alternative<bigtable::Row>(variant)) {
+      return absl::get<bigtable::Row>(std::move(variant));
+    }
+
+    auto status = absl::get<Status>(std::move(variant));
     if (status.ok()) {
-      return row;
+      return Status();
```
Comment (coryan) [inline google/cloud/bigtable/internal/legacy_row_reader_impl.cc:120]: nit: nowadays we would write `if (status.ok()) return Status{};`

Comment (coryan) [reply google/cloud/bigtable/internal/legacy_row_reader_impl.cc:120]: Here and some places below.

File: google/cloud/bigtable/internal/rowreaderiterator.cc [hunk 2]
```diff
@@ -41,14 +41,14 @@ RowReaderIterator& RowReaderIterator::operator++() {
 }
 
 void RowReaderIterator::Advance() {
-  auto status_or_optional_row = owner_->Advance();
-  if (!status_or_optional_row) {
-    row_ = StatusOr<bigtable::Row>(std::move(status_or_optional_row).status());
+  auto variant = owner_->Advance();
+  if (absl::holds_alternative<bigtable::Row>(variant)) {
```
Comment (coryan) [inline google/cloud/bigtable/internal/rowreaderiterator.cc:45]: You could do this with a visitor (but double check my ref-ref):

```cc
  struct Visitor {
    StatusOr<bigtable::Row> operator()(Status&& s) { return s; }
    StatusOr<bigtable::Row> operator()(bigtable::Row&& r) { return r; }
  };
  row_ = absl::visit(Visitor{}, owner_->Advance());
}
```

**Review by coryan — APPROVED** [review]

File: google/cloud/bigtable/internal/legacy_row_reader_impl.cc (hunk 1 above)
Comment (dbolduc) [reply google/cloud/bigtable/internal/legacy_row_reader_impl.cc]: Fixed.

File: google/cloud/bigtable/internal/rowreaderiterator.cc (hunk 2 above)
Comment (dbolduc) [reply google/cloud/bigtable/internal/rowreaderiterator.cc:45]: Won't Fix. I am going to delete the class shortly.

Also... what happens if your code is given an OK Status? 🤔

**Pull request merged**